tempesta_fw/tls.c

```diff
@@ -22,319 +22,548 @@
 #include "cfg.h"
 #include "connection.h"
 #include "client.h"
+#include "msg.h"
+#include "procfs.h"
 #include "tls.h"
 
-#define tls_dbg(c, fmt, ...)					\
-	TFW_DBG("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
-#define tls_err(c, fmt, ...)					\
-	TFW_ERR("TLS(%p,%p):%s " fmt "\n", c, c->sk, __func__, ## __VA_ARGS__)
+static struct {
+	ttls_config	cfg;
+	ttls_x509_crt	crt;
+	ttls_pk_context	key;
+	unsigned long	crt_pg_addr;
+	unsigned int	crt_pg_order;
+} tfw_tls;
 
-#define MAX_TLS_PAYLOAD		1460
-
-typedef struct {
-	mbedtls_ssl_config	cfg;
-	mbedtls_x509_crt	crt;
-	mbedtls_pk_context	key;
-} TfwTls;
-
-static TfwTls tfw_tls;
-
-static inline int
-ttls_ssl_handshake(TfwTlsContext *tls)
+/**
+ * Chop skb list with begin at @skb by TLS extra data at the begin and end of
+ * the the list after decryption and write the right pointer at the first skb
+ * and offset to @data for upper layers processing.
+ */
+static int
+tfw_tls_chop_skb_rec(TlsCtx *tls, struct sk_buff *skb, unsigned int off,
+		     TfwFsmData *data)
 {
-	int r;
+	size_t prolog = ttls_payload_off(&tls->xfrm);
+
+eaten_skb:
+	if (likely(skb->len > prolog)) {
+		off += prolog;
+	} else {
+		struct sk_buff *next = skb->next;
+		if (WARN_ON_ONCE(!next))
+			return -EIO;
+		prolog -= skb->len;
+		off = 0; /* off is only for the first skb */
+		__kfree_skb(skb);
+		skb = next;
+		goto eaten_skb;
+	}
+	tls->io_in.skb_list = NULL; /* not used any more */
 
-	spin_lock(&tls->lock);
-	r = mbedtls_ssl_handshake(&tls->ssl);
-	spin_unlock(&tls->lock);
+	data->skb = skb;
+	data->off = off;
+	data->trail = ttls_xfrm_taglen(&tls->xfrm);
 
-	return r;
+	return 0;
 }
 
-static inline int
-ttls_ssl_read(TfwTlsContext *tls, unsigned char *data, size_t size)
+static int
+tfw_tls_msg_process(void *conn, TfwFsmData *data)
 {
-	int r;
-
-	spin_lock(&tls->lock);
-	r = mbedtls_ssl_read(&tls->ssl, data, size);
-	spin_unlock(&tls->lock);
-
-	return r;
-}
+	int r, parsed = 0;
+	struct sk_buff *nskb = NULL, *skb = data->skb;
+	unsigned int off = data->off;
+	TfwConn *c = conn;
+	TlsCtx *tls = tfw_tls_context(c);
+	struct sk_buff *msg_skb;
+	TfwFsmData data_up = {};
 
-static inline int
-ttls_ssl_write(TfwTlsContext *tls, const unsigned char *data, size_t size)
-{
-	int r;
+	BUG_ON(off >= skb->len);
 
+	/*
+	 * Perform TLS handshake if necessary and decrypt the TLS message
+	 * in-place by chunks. Add skb to the list to build scatterlist if it
+	 * it contains end of current message.
+	 */
 	spin_lock(&tls->lock);
-	r = mbedtls_ssl_write(&tls->ssl, data, size);
-	spin_unlock(&tls->lock);
+next_msg:
+	ss_skb_queue_tail(&tls->io_in.skb_list, skb);
+	msg_skb = tls->io_in.skb_list;
```

Comment:
I guess, `skb_list` processing can be made without `msg_skb` variable: here and in `tfw_tls_chop_skb_rec()` - we have access to `tls->io_in.skb_list`.

Reply:
Yeah, now we modify `tls->io_in.skb_list` in `tfw_tls_chop_skb_rec()`: there was a bug that `tfw_tls_msg_process()` gets `off` from TCP layer as an overlap of sequence numbers, but we use the offset as TLS header for HTTP processing.

Reply:
Actually ttls_recv() initializes IO context after each processed TLS record so tfw_tls_msg_process() must store skb_list for further chopping in tfw_tls_chop_skb_rec(). Since ttls_recv() zeroes almost full IO, tfw_tls_chop_skb_rec() doesn't need to zero it. Fixed.